File: ruby/proverb/proverb.rb
class Proverb
  def initialize(*chain, qualifier: '')
    @chain = chain
    @qualifier = qualifier
  end

  def to_s
    main + final
  end

  private

  def main
    @chain.each_cons(2).each_with_object('') do |(word, next_word), str|
      str << "For want of a #{word} the #{next_word} was lost.\n"
    end
  end

  def final
    "And all for the want of a #{@qualifier} #{@chain.first}.".squeeze(' ')
  end
end

File: typescript/proverb/proverb.ts
export function proverb(...words: string[]): string {
  return words.reduce((lines, word, idx) => {
    if(words.length - 1 > idx) {
      return lines + standardLine(word, words[idx + 1])
    } 
    return lines + lastLine(words[0])
  }, "")
}

function standardLine(want: string, lost: string): string {
  return `For want of a ${want} the ${lost} was lost.\n`
}

function lastLine(initial: string): string {
  return `And all for the want of a ${initial}.`
}
